# Model Algorithm 

#### Linear Regressor
#### K Neighbors Regressor
#### Decision Tree Regressor
#### Gradient boosting Regressor



In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import warnings
warnings.filterwarnings('ignore')


In [2]:
df= pd.read_csv(r'C:\Users\user\Downloads\Scaled_Upper_Secondary_edu.csv')

In [3]:
df.head()

Year  Completion Rate for upper secondary Edu  Childhood Education GER  \
0  2012                                     96.9                47.348558   
1  2012                                     96.2                47.348558   
2  2013                                     97.0                47.348558   
3  2013                                     96.4                47.348558   
4  2014                                     97.2                36.650000   

   Gross enrolment ratio, early childhood educational development programmes  \
0                                           5.871662                           
1                                           5.871662                           
2                                           5.871662                           
3                                           5.871662                           
4                                           0.000000                           

   Literacy rate for 25-64 years old  \
0                             68.735   
1                             68.735   
2                             68.735   
3                             68.735   
4                             68.735   

   Expenditure on education as a percentage of total government expenditure (%)  \
0                                              19.25                              
1                                              19.25                              
2                                              17.96                              
3                                              17.96                              
4                                              17.32                              

   Government expenditure on education as a percentage of GDP (%)  \
0                                               3.90                
1                                               3.90                
2                                               3.44                
3                                               3.44                
4                                               3.42                

   Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
0           1.0                        0.0                             0.0   
1           1.0                        0.0                             0.0   
2           1.0                        0.0                             0.0   
3           1.0                        0.0                             0.0   
4           1.0                        0.0                             0.0   

   Europe and Northern America  Latin America and the Caribbean  \
0                          0.0                              0.0   
1                          0.0                              0.0   
2                          0.0                              0.0   
3                          0.0                              0.0   
4                          0.0                              0.0   

   Northern Africa and Western Asia  Oceania  Southern Asia  \
0                               0.0      0.0            0.0   
1                               0.0      0.0            0.0   
2                               0.0      0.0            0.0   
3                               0.0      0.0            0.0   
4                               0.0      0.0            0.0   

   Sub-Saharan Africa  Gender_numerical  
0                 0.0                 1  
1                 0.0                 0  
2                 0.0                 1  
3                 0.0                 0  
4                 0.0                 1

## Data Transformation

#### 1. Data Rescaling

In [4]:
# Data Rescaling

data_col = df['Year']
df.set_index('Year',inplace =True)

df.head()


Completion Rate for upper secondary Edu  Childhood Education GER  \
Year                                                                     
2012                                     96.9                47.348558   
2012                                     96.2                47.348558   
2013                                     97.0                47.348558   
2013                                     96.4                47.348558   
2014                                     97.2                36.650000   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2012                                           5.871662                           
2012                                           5.871662                           
2013                                           5.871662                           
2013                                           5.871662                           
2014                                           0.000000                           

      Literacy rate for 25-64 years old  \
Year                                      
2012                             68.735   
2012                             68.735   
2013                             68.735   
2013                             68.735   
2014                             68.735   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                              19.25                              
2012                                              19.25                              
2013                                              17.96                              
2013                                              17.96                              
2014                                              17.32                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                               3.90                
2012                                               3.90                
2013                                               3.44                
2013                                               3.44                
2014                                               3.42                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                              0.0   
2013                          0.0                              0.0   
2014                          0.0                              0.0   

      Northern Africa and Western Asia  Oceania  Southern Asia  \
Year                                                             
2012                               0.0      0.0            0.0   
2012                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   
2014                               0.0  

In [5]:
scaler = MinMaxScaler()
normalized_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)
normalized_df['Year'] = data_col
normalized_df.set_index('Year', inplace =True)
normalized_df.head()

Completion Rate for upper secondary Edu  Childhood Education GER  \
Year                                                                     
2012                                 0.985625                 0.291174   
2012                                 0.978437                 0.291174   
2013                                 0.986652                 0.291174   
2013                                 0.980491                 0.291174   
2014                                 0.988705                 0.224366   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2012                                           0.062772                           
2012                                           0.062772                           
2013                                           0.062772                           
2013                                           0.062772                           
2014                                           0.000000                           

      Literacy rate for 25-64 years old  \
Year                                      
2012                           0.657782   
2012                           0.657782   
2013                           0.657782   
2013                           0.657782   
2014                           0.657782   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                           0.482729                              
2012                                           0.482729                              
2013                                           0.446206                              
2013                                           0.446206                              
2014                                           0.428086                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                           0.286765                
2012                                           0.286765                
2013                                           0.252941                
2013                                           0.252941                
2014                                           0.251471                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                              0.0   
2013                          0.0                              0.0   
2014                          0.0                              0.0   

      Northern Africa and Western Asia  Oceania  Southern Asia  \
Year                                                             
2012                               0.0      0.0            0.0   
2012                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   
2014                               0.0  

In [6]:
normalized_df.corr()

Completion Rate for upper secondary Edu  \
Completion Rate for upper secondary Edu                                            1.000000   
Childhood Education GER                                                            0.232376   
Gross enrolment ratio, early childhood educatio...                                 0.350079   
Literacy rate for 25-64 years old                                                  0.312730   
Expenditure on education as a percentage of tot...                                -0.055010   
Government expenditure on education as a percen...                                 0.100356   
Central Asia                                                                       0.250122   
Central and Southern Asia                                                          0.046776   
Eastern and South-Eastern Asia                                                     0.064900   
Europe and Northern America                                                        0.450880   
Latin America and the Caribbean                                                    0.176922   
Northern Africa and Western Asia                                                   0.088123   
Oceania                                                                            0.003803   
Southern Asia                                                                     -0.119431   
Sub-Saharan Africa                                                                -0.601039   
Gender_numerical                                                                   0.031859   

                                                    Childhood Education GER  \
Completion Rate for upper secondary Edu                            0.232376   
Childhood Education GER                                            1.000000   
Gross enrolment ratio, early childhood educatio...                 0.298705   
Literacy rate for 25-64 years old                                  0.118609   
Expenditure on education as a percentage of tot...                -0.182103   
Government expenditure on education as a percen...                -0.017314   
Central Asia                                                      -0.064456   
Central and Southern Asia                                          0.087568   
Eastern and South-Eastern Asia                                     0.157829   
Europe and Northern America                                        0.185646   
Latin America and the Caribbean                                    0.024161   
Northern Africa and Western Asia                                  -0.050692   
Oceania                                                            0.053348   
Southern Asia                                                      0.149808   
Sub-Saharan Africa                                                -0.346218   
Gender_numerical                                                   0.001945   

                                                    Gross enrolment ratio, early childhood educational development programmes  \
Completion Rate for upper secondary Edu                                                      0.350079                           
Childhood Education GER                                                                      0.298705                           
Gross enrolment ratio, early childhood educatio...                                           1.000000                           
Literacy rate for 25-64 years old                                                            0.122515                           
Expenditure on education as a percentage of tot...                                           0.002370                           
Government expenditure on education as a percen...                                          -0.017604                           
Central Asia                                                                                -0.008424                           
Central and Southern Asia                                   

In [7]:
# spliting the dataset into train set, test set, validation set.
datax = normalized_df.drop('Completion Rate for upper secondary Edu', axis = 1)
datay = normalized_df['Completion Rate for upper secondary Edu']

train_ratio = 0.75
val_ratio = 0.15
test_ratio = 0.10

# making the train set to be 75% of the entire dataset.
x_train, x_test, y_train, y_test = train_test_split(datax, datay, test_size = 1-train_ratio)

# test is now 10% of the entire dataset
# validation is now 15% of the entire dataset

x_val, x_test, y_val, y_test = train_test_split(x_test, y_test, test_size = test_ratio/(test_ratio + val_ratio))

print(x_train, x_val, x_test)

      Childhood Education GER  \
Year                            
2017                 0.143187   
2019                 0.173598   
2013                 0.291174   
2019                 0.291174   
2018                 0.291174   
...                       ...   
2013                 0.135257   
2015                 0.291174   
2015                 0.291174   
2013                 0.291174   
2017                 0.289622   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2017                                           0.000000                           
2019                                           0.062772                           
2013                                           0.062772                           
2019                                           0.058157                           
2018                                           0.062772                 

## Model Training, Testing and evaluation

 
 Because the target variable been 'Completion rate for primary school Edu' is a continoues data and not categorical. therfore, the model algorithm to be use is the regression and not classifiers.

#### 1. Linear Regression


In [8]:

import time

# Record start time
start_time = time.time()

# Create a Linear Regression model
linear_model = LinearRegression()

# Fit the model to the training dataset
linear_model.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))





Training took 0.14 seconds.


In [9]:
# Obtain predictions
predicted_values = linear_model.predict(x_test)

#Measuring regression performance
MAE = mean_absolute_error(y_test, predicted_values)
MSE = mean_squared_error(y_test, predicted_values)
RMSE =np.sqrt(mean_squared_error(y_test, predicted_values))
RSS =np.sum(np.square(y_test, predicted_values))
r2=r2_score(y_test, predicted_values)




In [10]:
# Model evaluation

results = pd.DataFrame([['Linear regression', MAE, MSE, RMSE, RSS,r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results

Model      MAE       MSE      RMSE        RSS  r2_score
0  Linear regression  0.14755  0.036914  0.192131  65.296463  0.576905

In [11]:
lreg_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":predicted_values})
lreg_df


actual_values  predictions
Year                            
2020       0.595749     0.354917
2020       0.341822     0.116842
2017       0.137488     0.018903
2013       0.794640     0.631453
2018       0.753260     0.567401
...             ...          ...
2016       0.231954     0.053803
2020       0.467091     0.218174
2017       0.876065     0.767490
2016       0.271999     0.073984
2012       0.071773     0.005151

[188 rows x 2 columns]

#### Decition Tree Regressor

In [12]:
from sklearn.tree import DecisionTreeRegressor

import time


# Record start time
start_time = time.time()

DecisionTree_model = DecisionTreeRegressor()
DecisionTree_model.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))


Training took 0.14 seconds.


In [13]:
y_preds = DecisionTree_model.predict(x_test)

In [14]:
# Model Evaluation
mae= mean_absolute_error(y_test, y_preds)
mse= mean_squared_error(y_test, y_preds)
rmse=np.sqrt(mean_squared_error(y_test, y_preds))
rss=np.sum(np.square(y_test, y_preds))
r_squared=r2_score(y_test, y_preds)

In [15]:
model_results = pd.DataFrame([['Decision Tree regression', mae, mse, rmse, rss,r_squared]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [16]:
model_results

Model      MAE       MSE      RMSE        RSS  r2_score
0  Decision Tree regression  0.07333  0.021152  0.145436  65.296463  0.576905

In [17]:
Dreg_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":predicted_values})
Dreg_df


actual_values  predictions
Year                            
2020       0.595749     0.354917
2020       0.341822     0.116842
2017       0.137488     0.018903
2013       0.794640     0.631453
2018       0.753260     0.567401
...             ...          ...
2016       0.231954     0.053803
2020       0.467091     0.218174
2017       0.876065     0.767490
2016       0.271999     0.073984
2012       0.071773     0.005151

[188 rows x 2 columns]

#### K-Nearest Neighbors Regression (KNN Regression)

In [18]:

# Record start time
start_time = time.time()

knn = KNeighborsRegressor()

knn.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))




Training took 0.13 seconds.


In [19]:
knn_predictions  = knn.predict(x_test)

In [20]:
print(knn_predictions.shape, y_test.shape)

(188,) (188,)


In [21]:
# getting model scores
knn_r2 = r2_score(y_test, knn_predictions)
knn_mse = mean_squared_error( y_test,knn_predictions)
knn_rmse = np.sqrt(knn_mse)
knn_mae = mean_absolute_error(y_test,knn_predictions)
knn_rss = np.sum(np.square(y_test, knn_predictions))

In [22]:
model_results = pd.DataFrame([['KNN regression', knn_mae, knn_mse, knn_rmse, knn_rss,knn_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [23]:
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  KNN regression  0.084294  0.017835  0.133547  65.296463  0.777553

In [24]:
knn_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":knn_predictions})
knn_df.sort_index(ascending=True, inplace=True)
knn_df

actual_values  predictions
Year                            
2012       0.071773     0.005151
2012       0.232981     0.054280
2012       0.087175     0.007600
2012       0.978437     0.957339
2012       0.933258     0.870971
...             ...          ...
2021       0.510011     0.260112
2021       0.957696     0.917181
2021       0.481466     0.231810
2021       0.543074     0.294930
2021       0.142212     0.020224

[188 rows x 2 columns]

In [25]:
from sklearn.ensemble import GradientBoostingRegressor


# Record start time
start_time = time.time()

Gbr = GradientBoostingRegressor()
Gbr.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))



Training took 0.45 seconds.


In [26]:
Gbr_predictions  = Gbr.predict(x_test)

In [27]:
# getting model scores
Gbr_r2 = r2_score(y_test, Gbr_predictions)
Gbr_mse = mean_squared_error( y_test,Gbr_predictions)
Gbr_rmse = np.sqrt(Gbr_mse)
Gbr_mae = mean_absolute_error(y_test,Gbr_predictions)
Gbr_rss = np.sum(np.square(y_test, Gbr_predictions))

In [28]:
model_results = pd.DataFrame([['GBR regression', knn_mae, knn_mse, knn_rmse, knn_rss,knn_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [29]:
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  GBR regression  0.084294  0.017835  0.133547  65.296463  0.777553

In [30]:
Gbr_df = pd.DataFrame(data ={"actual_values": y_test, "predictions":knn_predictions})
Gbr_df.sort_index(ascending=True, inplace=True)
Gbr_df

actual_values  predictions
Year                            
2012       0.071773     0.005151
2012       0.232981     0.054280
2012       0.087175     0.007600
2012       0.978437     0.957339
2012       0.933258     0.870971
...             ...          ...
2021       0.510011     0.260112
2021       0.957696     0.917181
2021       0.481466     0.231810
2021       0.543074     0.294930
2021       0.142212     0.020224

[188 rows x 2 columns]

In [31]:
# Printing the stored results of all the three models
results.drop_duplicates()


Model       MAE       MSE      RMSE        RSS  r2_score
0         Linear regression  0.147550  0.036914  0.192131  65.296463  0.576905
1  Decision Tree regression  0.073330  0.021152  0.145436  65.296463  0.576905
2            KNN regression  0.084294  0.017835  0.133547  65.296463  0.777553
3            GBR regression  0.084294  0.017835  0.133547  65.296463  0.777553

### conclusion

1.Linear Regression time for the model to train is 0.14 seconds

1.Decision Tree Regression seems to perform well in terms of minimizing errors and time for the model to tarin is 0.14 seconds.

2.KNN Regression and GBR Regression have higher R-squared values, suggesting better explanatory power.
also, the time for the two models to train are 0.13 and 0.45 respectively.

The Decision Tree, KNN, and GBR regression models seem to have better performance based on the R-squared value compared to Linear Regression. This suggests that these models might be capturing more complex relationships in the data.